core/state/trie_prefetcher.go

```diff
@@ -293,348 +248,143 @@ type subfetcher struct {
-	to.taskLock.Lock()
-	defer to.taskLock.Unlock()
-
-	return to.skips
-}
-
-func (to *trieOrchestrator) enqueueTasks(tasks [][]byte) {
-	to.taskLock.Lock()
-	defer to.taskLock.Unlock()
-
-	if len(tasks) == 0 {
-		return
-	}
-
-	// Add tasks to [pendingTasks]
-	if !to.tasksAllowed {
-		to.skips += len(tasks)
-		return
-	}
-	to.processingTasks.Add(len(tasks))
-	to.pendingTasks = append(to.pendingTasks, tasks...)
-
-	// Wake up processor
-	select {
-	case to.wake <- struct{}{}:
-	default:
-	}
-}
-
-func (to *trieOrchestrator) handleStop(remaining int) {
-	to.taskLock.Lock()
-	to.skips += remaining
-	to.taskLock.Unlock()
-	to.processingTasks.Add(-remaining)
-}
-
-func (to *trieOrchestrator) processTasks() {
-	defer close(to.loopTerm)
-
+	// Trie opened successfully, keep prefetching items
 	for {
-		// Determine if we should process or exit
 		select {
-		case <-to.wake:
-		case <-to.stop:
-			return
-		}
-
-		// Get current tasks
-		to.taskLock.Lock()
-		tasks := to.pendingTasks
-		to.pendingTasks = nil
-		to.taskLock.Unlock()
-
-		// Enqueue more work as soon as trie copies are available
-		lt := len(tasks)
-		for i := 0; i < lt; i++ {
-			// Try to stop as soon as possible, if channel is closed
-			remaining := lt - i
-			select {
-			case <-to.stop:
-				to.handleStop(remaining)
-				return
-			default:
-			}
-
-			// Try to create to get an active copy first (select is non-deterministic,
-			// so we may end up creating a new copy when we don't need to)
-			var t Trie
-			select {
-			case t = <-to.copyChan:
-			default:
-				// Wait for an available copy or create one, if we weren't
-				// able to get a previously created copy
+		case <-sf.wake:
+			// Subfetcher was woken up, retrieve any tasks to avoid spinning the lock
+			sf.lock.Lock()
+			tasks := sf.tasks
+			sf.tasks = nil
+			sf.lock.Unlock()
+
+			// Prefetch any tasks until the loop is interrupted
+			for i, task := range tasks {
 				select {
-				case <-to.stop:
-					to.handleStop(remaining)
+				case <-sf.stop:
+					// If termination is requested, add any leftover back and return
+					sf.lock.Lock()
+					sf.tasks = append(sf.tasks, tasks[i:]...)
+					sf.lock.Unlock()
 					return
-				case t = <-to.copyChan:
-				case to.copySpawner <- struct{}{}:
-					to.copies++
-					t = to.copyBase()
-				}
-			}
 
-			// Enqueue work, unless stopped.
-			fTask := tasks[i]
-			f := func() {
-				// Perform task
-				var err error
-				if len(fTask) == common.AddressLength {
-					_, err = t.GetAccount(common.BytesToAddress(fTask))
-				} else {
-					_, err = t.GetStorage(to.sf.addr, fTask)
+				case ch := <-sf.copy:
+					// Somebody wants a copy of the current trie, grant them
+					ch <- sf.db.CopyTrie(sf.trie)
+
+				default:
+					// No termination request yet, prefetch the next entry
+					if _, ok := sf.seen[string(task)]; ok {
+						sf.dups++
+					} else {
+						if len(task) == common.AddressLength {
+							sf.trie.GetAccount(common.BytesToAddress(task))
```

Comment:
we could do something more explicit here instead of wrapping GetAccount, if desired.